src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala

```diff
@@ -248,6 +248,30 @@ class SamRecordClipper(val mode: ClippingMode, val autoClipAttributes: Boolean)
     else clipEndOfRead(rec, numberOfBasesToClip)
   }
 
+  /** Converts all soft-clipping to hard-clipping.  Does not convert unless the clipping mode is [[ClippingMode.Hard]].
+    * Calculates any clipping required and delegates to [[clipStartOfRead()]] and [[clipEndOfRead()]].
+    * @param rec the record to be clipped
+    * @return the number of bases converted to hard-clipping from soft-clipping at the start and end of the read
+    *         respectively.
+    */
+  def toHardClip(rec: SamRecord): (Int, Int) = {
+    if (rec.unmapped || this.mode != ClippingMode.Hard) (0, 0) else {
+      val numBasesClippedStart = {
+        val clippingElems = rec.cigar.elems.takeWhile(_.operator.isClipping)
+        val numSoft       = clippingElems.filter(_.operator == Op.S).map(_.length).sum
+        if (numSoft > 0) this.clipStartOfRead(rec, clippingElems.map(_.length).sum)
+        numSoft
+      }
+      val numBasesClippedEnd = {
+        val clippingElems = rec.cigar.reverseIterator.takeWhile(_.operator.isClipping).toSeq
+        val numSoft       = clippingElems.filter(_.operator == Op.S).map(_.length).sum
```

Comment:
@tfenne I would be curious about your opinion in regards to calling map/filter/sum on `IndexedSeq` relative to performance.  In your experience, is it better to convert them to iterators or streams?

Reply:
I've started to tend towards `fold(0)(_ + _.length)` vs. `.map().sum`.  I don't have any solid empirical evidence that it's faster, but since it won't create any intermediate iterators or collections, I think it should be.  Then again, these are all 0-2 element things to being with so probably no big deal.